dex/networks/eth/contracts/ETHSwapV1.sol

```diff
@@ -0,0 +1,226 @@
+        address participant;
+        uint64 value;
+    }
+
+    // contractKey generates a key hash which commits to the contract data. The
+    // generated hash is used as a key in the swaps map.
+    function contractKey(Contract calldata c) public pure returns (bytes32) {
+        return sha256(bytes.concat(c.secretHash, bytes20(c.initiator), bytes20(c.participant), bytes8(c.value), bytes8(c.refundTimestamp)));
+    }
+
+    // Redemption is the information necessary to redeem a Contract. Since we
+    // don't store the Contract itself, it must be provided as part of the
+    // redemption.
+    struct Redemption {
+        Contract c;
+        bytes32 secret;
+    }
+
+    function secretValidates(bytes32 secret, bytes32 secretHash) public pure returns (bool) {
+        return sha256(bytes.concat(secret)) == secretHash;
+    }
+
+    // constructor is empty. This contract has no connection to the original
+    // sender after deployed. It can only be interacted with by users
+    // initiating, redeeming, and refunding swaps.
+    constructor() {}
+
+    // senderIsOrigin ensures that this contract cannot be used by other
+    // contracts, which reduces possible attack vectors.
+    modifier senderIsOrigin() {
+        require(tx.origin == msg.sender, "sender != origin");
+        _;
+    }
+
+    // retrieveRecord retrieves the current swap record for the contract.
+    function retrieveRecord(Contract calldata c)
+        private view returns (bytes32, bytes32, uint256)
+    {
+        bytes32 k = contractKey(c);
+        bytes32 record = swaps[k];
+        return (k, record, uint256(record));
+    }
+
+    // state returns the current state of the swap.
+    function state(Contract calldata c)
+        public view returns(Record memory)
+    {
+        (, bytes32 record, uint256 blockNum) = retrieveRecord(c);
+        Record memory r;
+        if (blockNum == 0) {
+            r.state = State.Empty;
+        } else if (record == RefundRecord) {
+            r.state = State.Refunded;
+        } else if (secretValidates(record, c.secretHash)) {
+            r.state = State.Redeemed;
+            r.secret = record;
+        } else {
+            r.state = State.Filled;
+            r.blockNumber =  blockNum;
+        }
+        return r;
+    }
+
+    // initiate initiates an array of Contracts.
+    function initiate(Contract[] calldata contracts)
+        public
+        payable
+        senderIsOrigin()
+    {
+        uint initVal = 0;
+        for (uint i = 0; i < contracts.length; i++) {
+            Contract calldata c = contracts[i];
+
+            require(c.value > 0, "0 val");
+            require(c.refundTimestamp > 0, "0 refundTimestamp");
+
+            bytes32 k = contractKey(c);
+            bytes32 record = swaps[k];
+            require(record == bytes32(0), "swap not empty");
+
+            record = bytes32(block.number);
+            require(!secretValidates(record, c.secretHash), "hash collision");
```

Comment:
Although just as impossible as this collision, the secret can also not be `RefundRecord`. You could find out what that hashes to and do a straight != with the constant and record if you wanted to.